### ISSUE TITLE
Question : user data dir / cache with Chrome headless?

### ISSUE THREAD

**Issue by eric-gagnon** [prerender/prerender#632]
Hello,

Not much a of issue, just more of general question to get a better understanding of how things are working together.

I am using a self-hosted version on Rhel7. Installation was easy aside some minor issue with Apache middleware (Prerender is installed on the same server that is serving our app).

I did some testing and from the render time it looks like the render is alway redowloading all the app ressources (in my case a angular application). It likely confirmed with the flags that are used to start Chrome (no user data dir specified by exemple).

Full page load with user interactive Chrome takes 781 ms with primed cache (using agressive cache headers and leveraging Angular build hash file name). With prerender I get 1.78 s

It's not a big issue for our first scenario (prerender for metadata validation from Facebook and Twitter) but we are considering using prerender for Google indexing all our pages with a max possible case of around 1 million pages, so then every optimisation count at least to get the first indexing compled (even more considering that Googlebot is not doing the job in one full-time pass). I do understand that this optimisation would only be worth for the initial prerender and less of an issue with a cache mecanism.

I guess I could play with the parameters in our installation. We will not host for external use, just for our app, I guess this could be a more of a issue with a high-volume shared service.

Any other non-obvious issue that I should consider?

Thanks for the great project!

Eric

**Comment by thoop** [comment]
Hi there! We recommending caching pages heavily and then recaching those pages to keep them up to date in order to serve most of your pages from your cache. That way your pages are mostly served from the cache with the recaching process getting the latest information to save into the cache, so the rendering time wouldn't affect much.

We start a new incognito context for each tab so they don't share any cache between the pages.

We've done some testing on creating a sharable cache between incognito contexts using a Prerender plugin but we haven't released anything publicly on that front yet.

Basically, heavy caching is your best bet currently.

Check out our ssr-test repo for some optimizations on things like creating tabs ahead of time to improve render time.

**Comment by eric-gagnon** [comment]
Hello,

Took a long time to follow, just have it in "production" now but for a low use scenario (support for twitterbot and facebookexternalhit) so I have the time to iron things out before movint to indexing.

I do understand about caching the rendered page but my issue is more the caching of the html files so they are not downloaded by chrome headless each time.

I was able to figure it out:

chromeFlags: ['--headless', '--disable-gpu', '--remote-debugging-port=9222', '--hide-scrollbars', '--disk-cache-dir=/opt/prerender/chrome_disk_cache']

The folder have the permission of the service user.

I do see the folder getting filed but the performance I see from the logs is not suggesting a major improvement.

I have added a new request recently that could be explaining the delay (I'm around 2.5 seconds render time) but I'll check some more in cas it something else related to how the tab instances are managed.

Side note : I had some trouble with the Apache configs (handling of query parameters) but was able to figure it out too. Using prerenderhost/https://mypagetorender?myparams instead of prerenderhost?render=https://mypagetorender?myparams helped too (I suspect a conflict there because of the double ?).

**Comment by eric-gagnon** [comment]
Oh, I was the problem, it's working very fine. I was using a page that was doing a search request, the render was more bound by the request time.

If I use a page with smaller / faster request (with is the main type to be indexed) I see the benefit.
I stopped the service, cleared the chrome disk cache folder and tested again:

// First access (download the files in cache)
2020-06-15T14:41:58.166Z got 200 in **3822ms**

// Second access (or any other pages of same type), it's visibly quicker:
2020-06-15T14:42:04.388Z got 200 in **1807ms**

**Comment by eric-gagnon** [comment]
Issue is closed but a quick follow-up if this can be usefull for someone else.
Things looks good so far, I have started google indexing. Have found a few tweaks like blocking the reeources download options (wich also prevent collecting google analytics when prerender render the page). I had to start with a minimal solution and I was getting around 100 pates indexed a day, page render where too slow.

I checked the cache memory plugin, the file cache (file or binary) version from cache manager looked promising (without having to install other technology stack) but reading the code made me realise it was not exaclty the behavior I was looking for. I changed the behavior of the memory cache plugin to add compression (angular html output file are huge (around 400ko, mostly because of css) and wrote a simple prefill tool in nodejs using axios and another axios lib to throttle the request and not kill the server. Took around a hour to pre-cache-prerender our starting 5000 pages set. With compression memory usage is fine for this small set and could do a full nightly refill as needed but aiming for 1 millions doc will require something else (including not having to rerender on server/service restart)..

The pre-cached-prerendered version obviously return very fast. I saw a nice improvement in indexed page today (380 pages instead of the average 100 pages over the previous days.

So prefilling the cache is an important piece but the strategy would be different depending on the site.

It's the first time I'm working on google indexing so learned along the way. Google doesnt respect nofollow and was going in alot of pages (and triggering the prerender) that where inside the pages but not listed in the sitemap. I removed the nofollow from the other pages so that I get the indexed pages as google go through it anyway...

But then the problem (for me) is that's harder to know where google will be going in the one million set so I'll probably go with full pre-cache / pre-render. I don't have the possibility to multiply server instances and the server was running at 100% when handling 3 prerender request per seconds. I saw alot of chrome instance getting created and killed, I will have to double check that (pre-creation of tab also? like it was suggested in the thread).
